text/0000-const-generics.md

```diff
@@ -0,0 +1,287 @@
+are invalid:
+
+```rust
+fn foo<const X: usize>() {
+    const Y: usize = X * 2;
+    static Z: (usize, usize)= (X, X);
+
+    struct Foo([i32; X]);
+}
+```
+
+This restriction can be analogized to the restriction on using type variables
+in types constructed in the body of functions - all of these declarations,
+though private to this item, must be independent of it, and do not have any
+of its parameters in scope.
+
+## Theory of equality for type equality of two consts
+
+During unification and the overlap check, it is essential to determine when two
+types are equivalent or not. Because types can now be dependent on consts, we
+must define how we will compare the equality of two constant expressions.
+
+For most cases, the equality of two consts follows the same reasoning you would
+expect - two constant values are equal if they are equal to one another. But
+there are some particular caveats.
+
+### Structural equality
+
+Const equality is determined according to the definition of structural equality
+defined in [RFC 1445][1445]. Only types which  have the "structural match"
+property can be used as const parameters. This would exclude floats, for
+example.
+
+The structural match property is intended as a stopgap until a final solution
+for matching against consts has been arrived at. It is important for the
+purposes of type equality that whatever solution const parameters use will
+guarantee that the equality is *reflexive*, so that a type is always the same
+type as itself. (The standard definition of equality for floating point numbers
+is not reflexive.)
+
+This may diverse someday from the definition used by match; it is not necessary
+that matching and const parameters use the same definition of equality, but the
+definition of equality used by match today is good enough for our purposes.
+
+Because consts must have the structural match property, and this property
+cannot be enforced for a type variable, it is not possible to introduce a const
+parameter which is ascribed to a type variable (`<T, const N: T>` is not
+valid)>
+
+### Equality of two abstract const expressions
+
+When comparing the equality of two abstract const expressions (that is, those
+that depend on a variable) we cannot compare the equality of their values
+because their values are determined by an const variable, the value of which is
+unknown prior to monomorphization.
+
+For this reason we will (initially, at least) treat the return value of const
+expressions as *projections* - values determined by the input, but which are
+not themselves known. This is similar to how we treat associated types today.
+When comparing the evaluation of an abstract const expression - which we'll
+call a *const projection* - to another const of the same type, its equality is
+always unknown.
+
+Therefore we can neither unify nor guarantee the nonunification of any const
+projection with any other const unless they are *syntactically identical.* That
```

Comment:
And this is even despite the section in the RFC:

> Therefore we can neither unify nor guarantee the nonunification of any const projection with any other const unless they are syntactically identical. That is, because we require that const equality is reflexive, we know that {N + 1} is equal to {N + 1}, but we don't know whether or not it is equal to {N * 2} or even to N or {1 + N}.

In that case, what is the usefulness of this definition of equality for abstract const expressions?